Allow texture preview to zoom and popout in separate window

Arnklit opened the pull request:
This adds a button that opens the preview in a popup and adds zoom controls. Resolves https://github.com/godotengine/godot-proposals/issues/10160.

EDIT: Original video, before I added zoom, see video in comment for the current behavior

https://github.com/user-attachments/assets/55906e3d-8b80-4723-a775-4cb83dfac72b

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -286,6 +400,35 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
+	right_upper_corner_container->set_v_size_flags(Control::SIZE_SHRINK_BEGIN);
+	add_child(right_upper_corner_container);
+	zoom_out_button = memnew(Button);
+	zoom_out_button->connect("pressed", callable_mp(this, &TexturePreview::on_zoom_out_pressed));
```
```suggestion
	zoom_out_button->connect(SceneStringName(pressed), callable_mp(this, &TexturePreview::on_zoom_out_pressed));
```
For all uses of this signal

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -286,6 +400,35 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
+	zoom_reset_button = memnew(Button);
+	zoom_reset_button->connect("pressed", callable_mp(this, &TexturePreview::on_zoom_reset_pressed));
+	zoom_reset_button->set_flat(true);
+	zoom_in_button = memnew(Button);
```
```suggestion

	zoom_in_button = memnew(Button);
```

Arnklit commented:
Rebased and reworked with the new LOD selector. Ready for review.

https://github.com/user-attachments/assets/6b03d767-5eae-485b-a6ab-2f7b08589697

Arnklit committed:
Add texture poput and zoom to texture preview

Arnklit commented:
Rebased

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -243,7 +304,59 @@ void TexturePreview::on_selected_mipmap_changed(double p_value) {
+	popout_dialog->set_title("Texture Preview");
+
+	Vector2 popout_size = get_tree()->get_root()->get_size();
+	popout_size = MAX(Vector2(400, 300), popout_size * 0.5);
```
```suggestion
	popout_size = (popout_size * 0.5).max(Vector2(400, 300));
```
I assume, the current code is wrong and takes the maximum of the two, so if `popup_size * 0.5` is `(500, 100)` this would be set to `(500, 100)`

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -117,16 +157,37 @@ void TexturePreview::_notification(int p_what) {
+void TexturePreview::_update_pan() {
+	float zoom_clamp = (0.5 - 0.5 / zoom_level) * texture_display->get_size().height;
+	pan.x = CLAMP(pan.x, -zoom_clamp, zoom_clamp);
+	pan.y = CLAMP(pan.y, -zoom_clamp, zoom_clamp);
```
```suggestion
	pan = pan.clampf(-zoom_clamp, zoom_clamp);
```

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -243,7 +304,59 @@ void TexturePreview::on_selected_mipmap_changed(double p_value) {
+	popout_dialog->add_child(texture_preview_copy);
+
+	popout_dialog->connect("canceled", callable_mp(this, &TexturePreview::on_popout_closed).bind(popout_dialog));
+	popout_dialog->connect("confirmed", callable_mp(this, &TexturePreview::on_popout_closed).bind(popout_dialog));
```
```suggestion
	popout_dialog->connect(SceneStringname(confirmed), callable_mp(this, &TexturePreview::on_popout_closed).bind(popout_dialog));
```

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -273,10 +386,12 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
 	texture_display->set_expand_mode(TextureRect::EXPAND_IGNORE_SIZE);
 	texture_display->set_material(texture_material);
 	texture_display->set_instance_shader_parameter("u_channel_factors", Vector4(1, 1, 1, 1));
+	texture_display->connect("gui_input", callable_mp(this, &TexturePreview::_texture_display_gui_input));
```
```suggestion
	texture_display->connect(SceneStringName(gui_input), callable_mp(this, &TexturePreview::_texture_display_gui_input));
```

AThousandShips commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -290,29 +405,62 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
+	right_upper_corner_container->set_v_size_flags(Control::SIZE_SHRINK_BEGIN);
+	add_child(right_upper_corner_container);
+	zoom_out_button = memnew(Button);
+	zoom_out_button->connect("pressed", callable_mp(this, &TexturePreview::on_zoom_out_pressed));
```
```suggestion
	zoom_out_button->connect(SceneStringName(pressed), callable_mp(this, &TexturePreview::on_zoom_out_pressed));
```
For all uses of this signal

Calinou commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -290,29 +405,62 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
+	left_upper_corner_container->set_v_size_flags(Control::SIZE_SHRINK_BEGIN);
+	add_child(left_upper_corner_container);
+
+	// Add color channel selector at the bottom left if more than 1 channel is available.
```
I see the channel selector in the top-left corner, just like in the inspector:

```suggestion
	// Add color channel selector at the top-left corner if more than 1 channel is available.
```

Calinou reviewed (commented):
Tested locally (rebased on top of `master` 321b8c944fa32ee94eef44a56925105d7091be03), it works as expected.

Some feedback:

- The speed at which you pan doesn't match the mouse cursor movement, which feels odd. Panning is much faster than expected, especially at higher zoom levels.

- It should be possible to pan the view with the middle mouse button, for consistency with the 2D editor viewport.

- Zooming in at higher levels feels less impactful, as the zoom percentage is increased linearly (e.g. going from 400% to 410% is a much lower increase than going from 100% to 110%). I would look at how the 2D editor viewport handles zoom for inspiration, as it uses nonlinear scaling for scroll wheel zoom to ensure it feels consistent across zoom levels.

- At the minimum width, the mipmap and zoom indicators can overlap:

<img width="196" height="86" alt="Image" src="https://github.com/user-attachments/assets/75f69918-bea8-4fe9-ac46-6dba3741e4b6" />

This can also occur in the inspector if it's made narrow enough, especially if the channel indicator is also visible. This is roughly the minimum width required for them to not overlap:

<img width="326" height="48" alt="Image" src="https://github.com/user-attachments/assets/77780d76-1315-4486-bb20-279bd0b0f25c" />

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -98,6 +105,39 @@ TextureRect *TexturePreview::get_texture_display() {
 	return texture_display;
 }
 
+void TexturePreview::_texture_display_gui_input(const Ref<InputEvent> &p_event) {
```
You should use ViewPanner for panning, instead of implementing it manually.

QbieShay commented:
I think this is useful functionality. I'm particularly fond of the per-channel view.

I tested the PR, and I have some UX comments:

It's a bit hard to understand whether some channels are active or not. I suggest using the same scheme as the 3d view: 
<img width="154" height="103" alt="image" src="https://github.com/user-attachments/assets/cd6241eb-0d85-451f-8a43-6e306f18b28e" />

I would also suggest 2 more buttons for the channel: RGB, RGBA. RGBA is selected by default. RGB shows RGB without A
Clicking RGB is exclusive with RGBA and exclusive with per-component, while components stay non-exclusive with one another.

Would it be possible to use the whole space to pan and zoom? it feels very wasteful to have all that lateral space unused.

I would also label the MIP button. I would not completely disable it when there's no mipmap, i would just leave it there with disabled buttons. Alternatively, the up button is active but clicking it shows an error message that asks the user to enable mip if they want more mip.

Arnklit commented:
Thanks for the feedback @AThousandShips and @Calinou . I'll submit fixes for those issues soon. @KoBeWi I did not realise there was a helper class for exactly that, makes sense. I'll have a go at moving over to that.

@QbieShay I didn't add the channel selectors and mipmap selectors, they are already in master, I just had to adjust them a bit to accomodate the UI. And I had to make sure they worked in the popped out preview.

QbieShay commented:
What about the zooming rect using the whole space?

Arnklit commented:
> What about the zooming rect using the whole space?

Yeah I think that should be doable. I'm trying to rewrite it to use the ViewPanner now and that kind of changes the entire structure of the thing. I'll see if I can make that work.
